docs/proposals/trimming-friendly-registrar-proposal.md

```diff
@@ -0,0 +1,483 @@
+  - Frameworks
+  - Third-party static libraries
+  - Other generated Objective-C classes
+- After ILC produces the object file, collect the information about the Objective-C classes we need to include
+  - Scan the binary for undefined symbols which follow a specific naming convention which would include the Objective-C class name:
+  ```bash
+  $ nm -u aoted-app.o | grep '^_registrar__'
+  ```
+- Collect the Objective-C source files from the required Objective-C classes and their dependencies
+- Build the Objective-C source files into a second object file
+- Link the AOTed object file together with the second object file + all the frameworks and third-party libraries
+
+### Registrar endpoint naming convention
+
+The names of the registrar callbacks should contain the Objective-C class or category name we should include and it should
+be easy to reliably parse the endpoint name: `_registrar__<L>_<objc-name>_<N>_<sanitized-selector>`.
+  - The `L` would be the length of the `objc-name` string (could be encoded as dec, hex, base64)
+  - The `sanitized-selector` would be the selector where `:` symbols are replaced with `_`.
+  - The `N` is the index of the selector in an array of sorted selectors of the class/category
+    - this is a form of de-duplication to avoid name clashes caused by the sanitization (for example `some:_selector` vs
+      `some_:selector`)
+    - the `sanitized-selector` is there mostly for debugging purposes to appear in stack traces, the selector would be
+      unique even without it
+
+---
+
+## How to generate the registrar?
+
+I see three ways of generating the registrar:
+- Keep using custom linker steps.
+- Use a pre-trimming MSBuild task that's based on Mono.Cecil (independent of ILLink).
+- Use roslyn source generators.
+  - the generated and pre-built code will have to be distributed with the libraries (the same way the bgen output is
+    pre-built and distributed with libraries)
+
+### Option A: Custom linker steps
+
+We can modify the existing code that generates the managed static registrar to generate the trimmable variant. This could
+be useful to validate the idea and iron out the design before we put any effort into a proper implementation.
+
+PoC: https://github.com/xamarin/xamarin-macios/compare/main...simonrozsival:xamarin-macios:net8.0-poc-trimmable-msr
+
+### Option B: Pre-trimming + post-trimming MSBuild tasks
+
+We would modify the DLLs directly before the code is AOTed. Instead of ILLink and custom linker steps, we would create
+an independent MSBuild task to perform the weaving. This would include generating the `__Registrar_Callback__` nested
+classes with UCO endpoints, proxy classes for generic types, and generating any `[DynamicDependency]` or other attributes
+to express all inverse dependencies. At the same time, we would generate the Objective-C code and any information about
+native dependencies (frameworks, static libraries, other generated Objective-C source files).
```

Comment:
If we go with the option B and we have a dedicated task which uses Mono.Cecil to load/process/write the assemblies, would we still need or have a dependency on ILLink? 

I am just thinking if in this case we would have multiple build tasks needing to load/process/write assemblies which would affect the build time.

Reply:
The goal is specifically to get rid of the ILLink step altogether. We might bring some of the code from the custom linker steps into the pre-processing task, but we shouldn't definitely run all three tools - the new generator task, ILLink, and then ILC.

Reply:
Thanks for the clarification!